(Preceding snippet PR) Improved explanation of Wait/Ticker/Timeout/Threads/EventQueue (#1320)
ARMmbed/Handbook: docs/api/io/flow_control.md

We can use Blinky to explore flow control and task management in Arm Mbed OS applications. We'll look at automated actions first, then move on to handling user actions.

## Flow control for delayed actions

If we want to automatically blink an LED, we have four main techniques:

1. [Busy wait](#busy-wait)

> 1. [Timer, Ticker, Timeout](#ticker-timeout)

Author: ```suggestion
1. [Ticker, Timeout](#ticker-timeout)
```
(Timer measures time but does not provide any delay/callback)

Reviewer: Was it maybe added because you could use timer as follows:

```
READ CURRENT TIME FROM TIMER AND STORE IN START_TIME
IN A WHILE LOOP:
   IF READ CURRENT TIME FROM TIMER - START_TIME >= DELAY:
       PERFORM ACTION
       READ_CURRENT AND STORE IN START_TIME
```

Author: Checking a timer in a while loop isn't recommended is any case, so I'd rather leave it out. I think a timer is mainly for measuring time like benchmarking a procedure, etc. but that's outside the scope of this doc.

Reviewer: Oh yes. I just suggested an explanation why it was there in the first place.

1. [Thread](#thread)
1. [EventQueue](#eventqueue)

> All of those above accomplish delays but cater for different requirements for precision, efficiency and context.

Reviewer: All above techniques... Or All the above...

> All of those above accomplish delays but cater for different requirements for precision, efficiency and context.

Reviewer 2: ```suggestion
All of those above accomplish delays but cater to different requirements for precision, efficiency and context.
```

<span class="tips">**Tip:** You may want to read the [power optimization](../apis/platform-concepts.html) tutorial to understand how to achieve power savings. </span>

### Busy wait

> Busy wait is a method that blocks the processor for a period of time. The processor runs at full power for the duration of the wait:

Reviewer 2: ```suggestion
This is a method that blocks the processor for a while. The processor runs at full power for the duration of the wait:
```

Reviewer: I prefer what was presented by Lingkai in this PR. It is less ambiguous.

[![View code](https://www.mbed.com/embed/?url=https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Busy-Wait)](https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Busy-Wait/blob/v6.0/main.cpp)

Notice `wait_us()` - it is part of the [Wait API](../mbed-os-api-doxy/group__platform__wait__api.html) to busy wait a given number of microseconds.

> The Wait API is a precise and ISR-safe way to create short delays (nanoseconds to microseconds). However, we recommend not busy waiting much longer than that due to inefficiency.

Reviewer 3: Yet more precise than an `sleep_for` it is still _best effort_ and should not be relied on for that.

Author: I'll remove "precise"

Author: Done

> The Wait API is a precise and ISR-safe way to create short delays (nanoseconds to microseconds). However, we recommend not busy waiting much longer than that due to inefficiency.

Reviewer: ```suggestion
The Wait API is a precise and ISR-safe way to create nanoseconds to microseconds delays. However, we do not recommend busy waiting for longer delays.
```

> The Wait API is a precise and ISR-safe way to create short delays (nanoseconds to microseconds). However, we recommend not busy waiting much longer than that due to inefficiency.

Reviewer: That is grammatically incorrect.
What was proposed is correct.

> The following few techniques are better suited for longer waits (milliseconds to seconds) which apply to our example's use case.

Reviewer: ```suggestion
The following techniques are better suited for milliseconds to seconds delays which apply to our example's use case.
```

### Ticker, Timeout

> Tickers and Timeouts are a non-blocking, interrupt-based way of creating a time interval - your code continues to execute or sleep when there is nothing to do. The difference is that Tickers are recurring whereas Timeouts are one-off.

Reviewer 2: ```suggestion
Tickers and Timeouts are a non-blocking, creating a time interval execution in interrupt context - your code continues to execute or sleep when there is nothing to do. The difference is that Tickers are recurring where Timeouts are one-off.
```

> Tickers and Timeouts are a non-blocking, interrupt-based way of creating a time interval - your code continues to execute or sleep when there is nothing to do. The difference is that Tickers are recurring whereas Timeouts are one-off.

Reviewer: The proposition is grammatically incorrect.
From Lingkai changes, I would only replace `interrupt-based way` with `interrupt-based ways`.

Here is an example that uses a ticker object:

[![View code](https://www.mbed.com/embed/?url=https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Ticker)](https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Ticker/blob/v6.0/main.cpp)

> <span class="warnings"> **Warning:** A ticker/timeout's handler (`flip` in this example), like any ISR handlers, should return quickly and not use `printf` or APIs that are not intended for ISRs.</span>

Reviewer 3: I would clearly remind here that the handlers are run in the ISR context : 

```suggestion
<span class="warnings"> **Warning:** A ticker/timeout's handlers are executed in ISR context and thus, like any ISR handlers, should return quickly and not use `printf` or APIs that are not intended for ISRs.</span>
```

> If you don't need the precision of a high-frequency ticker or timeout, we recommend that you use LowPowerTicker or LowPowerTimeout instead. This allows the system to be put in deep sleep mode.

Reviewer: ```suggestion
If you don't need the precision of a high-frequency ticker or timeout, we recommend that you use LowPowerTicker or LowPowerTimeout instead. These allow the system to be put in deep sleep mode.
```

> If you don't need the precision of a high-frequency ticker or timeout, we recommend that you use LowPowerTicker or LowPowerTimeout instead. This allows the system to be put in deep sleep mode.

Reviewer 2: ```suggestion
If you don't need the precision of a high-frequency ticker or timeout, we recommend LowPowerTicker or LowPowerTimeout instead to use. This allows the system to be put in deep sleep mode.
```

> If you don't need the precision of a high-frequency ticker or timeout, we recommend that you use LowPowerTicker or LowPowerTimeout instead. This allows the system to be put in deep sleep mode.

Reviewer: That does not read as well as what Lingkai suggested.

- _High-resolution microsecond ticker/timeout_ ([Ticker](ticker.html), [Timeout](timeout.html))
- _Low Power ticker/timeout_ ([LowePowerTicker](lowpowerticker.html), [LowePowerTimeout](lowpowertimeout.html))

> Usage of the low power classes will inform the operating system of your desire to enable Deep Sleep mode in your board, although enabling this depends on the specific environment and characteristics where your code runs. But it's helpful to know that there are options to enable this. For more information about Sleep and Deep Sleep, we would suggest to look at our documentation page [here](power-management-sleep.html) as well as watch our [Mbed Office Hours video](https://www.youtube.com/watch?v=OFfOlBaegdg&t=12s) where the concepts of Sleep and Deep Sleep are described in-depth.

Reviewer 3: > But it's helpful to know that there are options to enable this. 

This sounds/reads a bit too informal.

```suggestion
Usage of the low power classes will inform the operating system of your desire to allow Deep Sleep mode on your system, although actually entering deep sleep depends on the specific environment and characteristics of your system. For more information about Sleep and Deep Sleep, please refer to our [documentation page](power-management-sleep.html) as well as watch our [Mbed Office Hours video](https://www.youtube.com/watch?v=OFfOlBaegdg&t=12s) where the concepts of Sleep and Deep Sleep are described in-depth.
```

> Usage of the low power classes will inform the operating system of your desire to enable Deep Sleep mode in your board, although enabling this depends on the specific environment and characteristics where your code runs. But it's helpful to know that there are options to enable this. For more information about Sleep and Deep Sleep, we would suggest to look at our documentation page [here](power-management-sleep.html) as well as watch our [Mbed Office Hours video](https://www.youtube.com/watch?v=OFfOlBaegdg&t=12s) where the concepts of Sleep and Deep Sleep are descibed in-depth.

Reviewer 2: ```descibed => described ```

### Thread

If your application is running in RTOS mode then threads are another efficient way to blink an LED. During the waiting period, it is possible to take advantage of Mbed OS optimizations to automatically conserve power and deal with other tasks.

[![View code](https://www.mbed.com/embed/?url=https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Thread)](https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-Thread/blob/v6.0/main.cpp)

> Threads are the most efficient way to run tasks in Mbed OS since the operating system internally contains optimizations for efficiently scheduling the tasks and putting the device as much as possible to sleep (or deep sleep).

Reviewer: ```suggestion
Threads are the most efficient ways to run tasks in Mbed OS since the operating system internally contains optimizations for efficiently scheduling the tasks and putting the device as much as possible to sleep (or deep sleep).
```

> Threads are the most efficient way to run tasks in Mbed OS since the operating system internally contains optimizations for efficiently scheduling the tasks and putting the device as much as possible to sleep (or deep sleep).

Reviewer 3: Instead of

> and putting the device as much as possible to sleep (or deep sleep).

I would write

> and maximizing the system's sleep time.

Author: Done

### EventQueue

The EventQueue class has a simple-to-use `call_every()` function to schedule repeated actions:

> [![View code](https://www.mbed.com/embed/?url=https://github.com/ARMmbed/mbed-os-examples-docs_only/blob/master/Tutorials_UsingAPIs/Flow-Control-EventQueue)](https://github.com/ARMmbed/mbed-os-examples-docs_only/blob/master/Tutorials_UsingAPIs/Flow-Control-EventQueue/main.cpp)

Reviewer: I believe you need to update the URL to point to `https://github.com/ARMmbed/mbed-os-snippet-Flow-Control-EventQueue`.

Author: The sub repos of individual examples are currently only mirrored during releases. We haven't fully decided how to proceed yet.

> For one-off delays, `call_in()` is available. Just as with Ticker/Timeout, if no threads are running during a wait, the system enters sleep mode.

Reviewer 3: This also has the advantage over `Ticker/Timeout` of running in a thread context so `Mutex::lock` are legal.
Blocking wait should still be avoided there since it would block the whole queue.

Author: Good point. I'll read about this.

Author: > Blocking wait should still be avoided there since it would block the whole queue.

Waiting in an EventQueue which we use for waiting is a bit complex for the purpose of this doc I guess 😅

> Alternatively, you can use a Ticker/Timeout to create delays (like the earlier example) but delegate the action (e.g. blinking) from an ISR to the EventQueue using `EventQueue::call()` to lift restrictions like `printf`. This strategy is explained in [The EventQueue API](../the-eventqueue-api.html) tutorial.

Reviewer: ```suggestion
Alternatively, you can use a Ticker/Timeout to create delays (like in the earlier example) but delegate the action (e.g. blinking) from an ISR to the EventQueue using `EventQueue::call()` to lift restrictions like `printf`. This strategy is explained in [The EventQueue API](../the-eventqueue-api.html) tutorial.
```

> Alternatively, you can use a Ticker/Timeout to create delays (like the earlier example) but delegate the action (e.g. blinking) from an ISR to the EventQueue using `EventQueue::call()` to lift restrictions like `printf`. This strategy is explained in [The EventQueue API](../the-eventqueue-api.html) tutorial.

Reviewer 3: I don't see any advantage from this use case over the pure equeue use case. And this has a rather heavy cognitive complexity as it involves more concepts/tools.

Author: Maybe the only exception is when we use EventQueue with a low power timer to save power. But still, using an EventQueue to exit ISRs quickly is a general strategy which we already [document elsewhere](https://os.mbed.com/docs/mbed-os/v5.15/tutorials/the-eventqueue-api.html), so we may not need to explain anything here.

Author: I removed this paragraph, will push a commit later

Author: Done

> For one-off delays, `call_in()` is available. Just as with Ticker/Timeout, if no threads are running during the wait period, the system enters sleep mode.

Reviewer 2: ```suggestion
For one-off delays, `call_in()` is available. Just as with Ticker/Timeout, if no threads are running during the waiting period, the system enters sleep mode.
```

Author: I'll change it to "during a wait", because "wait" as a noun is already a period
